### PR TITLE
Deployment test

### PR DESCRIPTION
**Pull request by vardeep-motor-ai**
---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207445569898407

### CODE REVIEW

**Comment by zeezou17** [comment]
> _ No description provided. _

Please provide description of the things you have added in the PR.

File: src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#include "postprocess.cpp"
```
Comment (zeezou17) [inline src/main.cpp:1]: Please add doxygen style comments on the header and functions

File: src/postprocess.cpp [hunk 2]
```diff
@@ -1,27 +1,41 @@
 
+#include "prints.cpp"
```
Comment (zeezou17) [inline src/postprocess.cpp:2]: Please add doxygen style comments on the header and functions

File: src/postprocess.cpp [hunk 3]
```diff
@@ -1,27 +1,41 @@
...
+  float conf_threshold = 0.2;
+  bool refine_vis = false;
+  float vis_thresh = 0.5;
+  AnchorMat proposals_arr;
+  bool is_test;
 
 public:
+  // NMS(py::list &proposals) { proposals_arr =
+  // convert_pylist_to_arr(proposals); }
+  NMS(AnchorMat proposals, int nms_thold, bool is_test_flag) {
```
Comment (zeezou17) [inline src/postprocess.cpp]: No .hpp files. Each cpp file must have a corresponding .hpp files which defines the structure of the file. 
The purpose of .hpp is to give a high level overview of the functions and the classes by declarations and the doxygen comments go there.
A .cpp file includes the defnintion and implementation of the corresponding functions and classes and adds more details if needed using docygen tags

File: src/postprocess.cpp (hunk 2 above)
Comment (zeezou17) [inline src/postprocess.cpp]: Please add doxygen style comments on the header and functions

File: src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+AnchorMat create_sample_input() {
+  AnchorMat anchor_vec;
```
Comment (zeezou17) [inline src/main.cpp]: Please add information what is each anchor vector and what is the collection of anchor vestor supposed to represent

File: src/main.cpp (hunk 1 above)
Comment (zeezou17) [inline src/main.cpp]: Please add doxygen style comments on the header and functions

File: tests/test_nms.cpp [hunk 5]
```diff
@@ -57,3 +57,148 @@ TEST(TestSoftmax, Hello) {
...
+    }
+  }
+  std::cout << out_vec.size() << std::endl;
+  return out_vec;
+}
+
+TEST(TestFullFunctional, Hello) {
+
+  AnchorMat input_anchors = read_anchors_from_file(
+      "/home/sandhu/project/LaneSeg/Anchor3DLane/proposals.txt");
```
Comment (zeezou17) [inline tests/test_nms.cpp]: No Ansolute path onle variable paths and it the propasals.txt is a relevant testcase then it should be pushed

File: tests/test_nms.cpp [hunk 6]
```diff
@@ -57,3 +57,148 @@ TEST(TestSoftmax, Hello) {
...
+  AnchorMat required_output = read_anchors_from_file(
+      "/home/sandhu/project/LaneSeg/Anchor3DLane/output.txt");
```
Comment (zeezou17) [inline tests/test_nms.cpp]: No absolute path please

File: src/utils.cpp [hunk 7]
```diff
@@ -170,13 +178,18 @@ std::vector<float> test_scores(const AnchorMat &proposals_arr) {
...
+                 bool is_test) {
   std::vector<int> anchor_inds = create_sequence_vector(0, num_anchors - 1);
   AnchorMat proposals_after_thresholding;
   std::vector<float> scores_after_thresholding;
   std::vector<int> anchor_inds_after_thresholding;
-  // std::vector<float> scores = compute_score(proposals_arr);
-  std::vector<float> scores = test_scores(proposals_arr);
+  std::vector<float> scores;
+  if (is_test) {
+    scores = test_scores(proposals_arr);
```
Comment (zeezou17) [inline src/utils.cpp]: Whats the difference between test scores and compute scores??

File: src/utils.cpp [hunk 8]
```diff
@@ -43,8 +43,8 @@ std::vector<T> get_items_from_indices(std::vector<T> data,
 // }
 
 std::vector<int> create_sequence_vector(int start, int end) {
```
Comment (zeezou17) [inline src/utils.cpp:23]: Please doxygen style commensts

File: tests/test_nms.cpp [hunk 9]
```diff
@@ -1,8 +1,9 @@
-#include <gtest/gtest.h>
-
 #include "../src/postprocess.cpp"
-// #include "../src/utils.cpp"
-
+#include <cmath>
```
Comment (zeezou17) [inline tests/test_nms.cpp:4]: Please add doxygen style comments on the header and functions

File: tests/test_nms.cpp [hunk 10]
```diff
@@ -13,8 +14,6 @@ AnchorMat create_sample_input() {
   Anchor anchor_5{0.3, 1.6, 1.1, 0.9, 0.3};
```
Comment (zeezou17) [inline tests/test_nms.cpp]: Please define what each anchors are as well

File: tests/test_nms.cpp [hunk 11]
```diff
@@ -13,8 +14,6 @@ AnchorMat create_sample_input() {
...
   Anchor anchor_6{0.4, 1.7, 1.0, 1.0, 0.2};
```
Comment (zeezou17) [inline tests/test_nms.cpp]: Please provide a documentation as to why the test cases where created how they were chosen and what they represent and do they completely test your function. You can write all of this as doxygen comments after each test case and explain what this tests. I need to know the underlying logic of why these testcases are enough to test the robustness of the function.

File: tests/test_nms.cpp [hunk 12]
```diff
@@ -57,3 +57,148 @@ TEST(TestSoftmax, Hello) {
...
+  for (int i = 0; i < accumulated_dis.size(); ++i) {
+    ASSERT_TRUE(floatCompare(accumulated_dis[i], expected_output[i]));
+  }
+  // ASSERT_EQ(req_out, input_order);
+}
+
+AnchorMat conf_testing() {
+  AnchorMat anchor_vec;
+
+  Anchor anchor_2{0.7, 1.3, 1.4, 0.6, 0.6};
```
Comment (zeezou17) [inline tests/test_nms.cpp]: Please add the motivation of why this test case was chosen and why its important

File: tests/test_nms.cpp [hunk 13]
```diff
@@ -57,3 +57,148 @@ TEST(TestSoftmax, Hello) {
...
+  NMS test_obj(test_input, 0.2, true);
+  std::vector<int> keep = test_obj.nms_3d(proposals, scores, 0.2);
+  std::vector<int> expected_out_indices{3, 1, 5};
+  ASSERT_EQ(keep.size(), expected_out_indices.size());
+  for (int i = 0; i < expected_out_indices.size(); ++i) {
+    ASSERT_EQ(expected_out_indices[i], keep[i]);
+  }
+}
+
+TEST(TestFilterRows, Hello) {
```
Comment (zeezou17) [inline tests/test_nms.cpp]: What is the Hello supposed to represent in every test case ??

File: tests/test_nms.cpp [hunk 14]
```diff
@@ -57,3 +57,148 @@ TEST(TestSoftmax, Hello) {
...
+  ASSERT_EQ(expected_ouput.size(), output.size());
+
+  for (int i = 0; i < expected_ouput.size(); ++i) {
+    for (int j = 0; j < expected_ouput[i].size(); ++j) {
+      ASSERT_TRUE(floatCompare(expected_ouput[i][j], output[i][j]));
+    }
+  }
+}
+
+std::vector<std::vector<float>> read_anchors_from_file(std::string path) {
```
Comment (zeezou17) [inline tests/test_nms.cpp]: Maybe this should be a function in utils and not in test file and this function itself should have a test case. Also doxygen comments please

**Review by zeezou17 — CHANGES_REQUESTED** [review]
Please make the necessary changes and if you have any questions let me know. Thanks

**Comment by zeezou17** [comment]
@vardeep-motor-ai as discussed adding an example for the doxygen style commenting which is followed in the stack. Refer [here](https://github.com/Motor-Ai/Motor/tree/perception/logging_doxygen_main/Ros_Historic_Core/src/Sense/src/app/plugins/ros2_ids)

File: src/postprocess.cpp (hunk 3 above)
Comment (vardeep-motor-ai) [reply src/postprocess.cpp]: True. I have added the corresponding .hpp files

File: src/postprocess.cpp (hunk 2 above)
Comment (vardeep-motor-ai) [reply src/postprocess.cpp]: Done. I have added the doxygen comments for both postprocess.hpp and utils.hpp

File: src/main.cpp (hunk 4 above)
Comment (vardeep-motor-ai) [reply src/main.cpp]: This was actually a sample script where I was writing test code. But I have added the Anchor information

File: tests/test_nms.cpp (hunk 5 above)
Comment (vardeep-motor-ai) [reply tests/test_nms.cpp]: Done

File: tests/test_nms.cpp (hunk 10 above)
Comment (vardeep-motor-ai) [reply tests/test_nms.cpp]: Done

File: tests/test_nms.cpp (hunk 13 above)
Comment (vardeep-motor-ai) [reply tests/test_nms.cpp]: Changed it

File: tests/test_nms.cpp (hunk 11 above)
Comment (vardeep-motor-ai) [reply tests/test_nms.cpp]: Do I have to write a documentation about it?

File: tests/test_nms.cpp (hunk 12 above)
Comment (vardeep-motor-ai) [reply tests/test_nms.cpp]: Done!

File: tests/test_nms.cpp (hunk 14 above)
Comment (vardeep-motor-ai) [reply tests/test_nms.cpp]: Done!

File: tests/test_nms.cpp (hunk 11 above)
Comment (zeezou17) [reply tests/test_nms.cpp]: The purpose of unit tests is to write test cases that make sure that the function will not break or give errors or wrong answers no matter what input is given as the arguments to the function. So, I just want a short description of why you chose the numbers in the Anchors as your testcases are they representative enough to guarantee that the unit testing is complete. The short description helps me or somebody who works on your code later evaluate if you have though of the test cases from every angle and have not missed anything. 

A small example is below
```
int add_natural_numbers(int a, int b)
    return a + b
```

If a negative number is given then this should not add them but without the safety checks this testcase fails and the new code then becomes

```
int add_natural_numbers(int a, int b)
    if (a >0 && b > 0)
          return a + b
   else
          throws excception("only positive numbers excepted")
```

File: include/postprocess.hpp [hunk 15]
```diff
@@ -0,0 +1,140 @@
+#ifndef POSTPROCESS_HPP
```
Comment (zeezou17) [inline include/postprocess.hpp:10]: Forgot to add the doxygen comments that ar top of the files in .hpp

File: include/utils.hpp [hunk 16]
```diff
@@ -0,0 +1,169 @@
+#ifndef UTILS_HPP
```
Comment (zeezou17) [inline include/utils.hpp:10]: Same file doxygen comments

**Review by zeezou17 — CHANGES_REQUESTED** [review]
Overall looks good lots of nice changes missed a few places and just one major change left from the previous review adding some information about the chose test cases I have also given an example

File: include/utils.hpp (hunk 16 above)
Comment (vardeep-motor-ai) [reply include/utils.hpp:10]: Done

File: include/postprocess.hpp (hunk 15 above)
Comment (vardeep-motor-ai) [reply include/postprocess.hpp:10]: Done

**Pull request merged**